org.eclipse.jdt.core/dom/org/eclipse/jdt/core/dom/ASTParser.java

```diff
@@ -612,6 +614,7 @@ public void setSource(char[] source) {
 	 */
 	public void setSource(ICompilationUnit source) {
 		setSource((ITypeRoot)source);
+		setCompilerOptions(source.getOptions(true));
```

Comment:
The test failures are because of this part. This needs a `null` check because :

```
@param source the Java model compilation unit whose source code
	 * is to be parsed, or <code>null</code> if none
```

The code formatter logic also has some code that attempts to parse code as various different kinds (cu, expression, body, statement, module, etc.) and when this new statement is now called, it fails because `source.getOptions(true)` triggers some hashcode logic that requires the CU owner, which is `null`.